src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -5546,5 +5509,330 @@ static void Main()
+
+            var x1 = GetDeconstructionVariable(tree, "x1");
+            var x1Ref = GetReference(tree, "x1");
+            VerifyModelForDeconstructionLocal(model, x1, x1Ref);
+
+            var x2 = GetDeconstructionVariable(tree, "x2");
+            var x2Ref = GetReference(tree, "x2");
+            VerifyModelForDeconstructionLocal(model, x2, x2Ref);
+        }
+
+        [Fact]
+        public void MixedDeconstruction_05()
+        {
+            string source = @"
+class Program
+{
+    static void Main(string[] args)
+    {
+        foreach ((M(out var x1), args is var x2, _) in new[] { (1, 2, 3) })
+        {
+            System.Console.WriteLine(x1);
+            System.Console.WriteLine(x2);
+        }
+    }
+    static int _M;
+    static ref int M(out int x) { x = 2; return ref _M; }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            compilation.VerifyDiagnostics(
+                // (6,34): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         foreach ((M(out var x1), args is var x2, _) in new[] { (1, 2, 3) })
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "args is var x2").WithLocation(6, 34),
+                // (6,34): error CS0029: Cannot implicitly convert type 'int' to 'bool'
+                //         foreach ((M(out var x1), args is var x2, _) in new[] { (1, 2, 3) })
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "args is var x2").WithArguments("int", "bool").WithLocation(6, 34),
+                // (6,18): error CS8186: A foreach loop must declare its iteration variables.
+                //         foreach ((M(out var x1), args is var x2, _) in new[] { (1, 2, 3) })
+                Diagnostic(ErrorCode.ERR_MustDeclareForeachIteration, "(M(out var x1), args is var x2, _)").WithLocation(6, 18)
+                );
+            var tree = compilation.SyntaxTrees.First();
+            var model = compilation.GetSemanticModel(tree);
+
+            var x1 = GetDeconstructionVariable(tree, "x1");
+            var x1Ref = GetReference(tree, "x1");
+            Assert.Equal("int", model.GetTypeInfo(x1Ref).Type.ToDisplayString());
+
+            model = compilation.GetSemanticModel(tree);
+            var x2 = GetDeconstructionVariable(tree, "x2");
+            var x2Ref = GetReference(tree, "x2");
+            Assert.Equal("string[]", model.GetTypeInfo(x2Ref).Type.ToDisplayString());
+
+            VerifyModelForLocal(model, x1, LocalDeclarationKind.RegularVariable, x1Ref);
+            VerifyModelForLocal(model, x2, LocalDeclarationKind.PatternVariable, x2Ref);
+        }
+
+        [Fact]
+        public void ForeachIntoExpression()
+        {
+            string source = @"
+class Program
+{
+    static void Main(string[] args)
+    {
+        foreach (M(out var x1) in new[] { 1, 2, 3 })
```

Comment:
See the very next test  here.

---
In reply to: [91582314](https://github.com/dotnet/roslyn/pull/15548#discussion_r91582314) [](ancestors = 91582314)